### PR TITLE
[XXXX] Make puma auto-generate SSL certificates

### PR DESCRIPTION
**Pull request by tvararu**
Removes one step from the initial setup.

### CODE REVIEW

File: config/puma.rb [hunk 1]
```diff
@@ -29,17 +30,35 @@
 # Allow puma to be restarted by `rails restart` command.
 plugin :tmp_restart
 
-# SSL in development (as DFE Signin redirects to https://localhost:3000 but
-# just listen on port 3000 elsewhere
 listen_port = ENV.fetch("PORT") { 3000 }
 
-if Rails.env.development?
+if env == "development"
```
Comment (dankmitchell) [inline config/puma.rb]: I think I prefer `if Rails.env.development?`

File: config/puma.rb [hunk 2]
```diff
@@ -9,7 +9,8 @@
 
 # Specifies the `environment` that Puma will run in.
 #
-environment ENV.fetch("RAILS_ENV") { "development" }
+env = ENV.fetch("RAILS_ENV") { "development" }
+environment env
```
Comment (dankmitchell) [inline config/puma.rb]: What's the benefit of this diff?

File: config/puma.rb [hunk 3]
```diff
@@ -29,17 +30,35 @@
...
 listen_port = ENV.fetch("PORT") { 3000 }
 
-if Rails.env.development?
+if env == "development"
+  cert = "#{Dir.pwd}/#{File.join('config', 'localhost', 'https', 'localhost.crt')}"
+  key = "#{Dir.pwd}/#{File.join('config', 'localhost', 'https', 'localhost.key')}"
 
-  cert = Rails.root.join('config', 'localhost', 'https', 'localhost.crt')
-  key = Rails.root.join('config', 'localhost', 'https', 'localhost.key')
+  unless File.exist?(key)
```
Comment (dankmitchell) [inline config/puma.rb]: Do we need to check for  `File.exist?(cert) && File.exist?(key)` ?

File: config/puma.rb (hunk 2 above)
Comment (tvararu) [reply config/puma.rb]: I added it because we re-use the env variable later on

File: config/puma.rb (hunk 1 above)
Comment (tvararu) [reply config/puma.rb]: We can't use `Rails` because we call `puma` directly in the procfile, it's weird...

File: config/puma.rb (hunk 3 above)
Comment (tvararu) [reply config/puma.rb]: Yeah may as well

File: config/puma.rb [hunk 4]
```diff
@@ -9,7 +9,8 @@
 
 # Specifies the `environment` that Puma will run in.
 #
-environment ENV.fetch("RAILS_ENV") { "development" }
+env = ENV.fetch("RAILS_ENV") { "development" }
```
Comment (misaka) [inline config/puma.rb]:
```
ENV.fetch('RAILS_ENV', 'development')
```

The block is just used when you don't want to automatically execute `"development"` every time it runs. For example what if that was `YAML.load(File.read(".config.yml"))[:environment]` ... you wouldn't want to execute that unless `RAILS_ENV` wasn't set.

File: config/puma.rb [hunk 5]
```diff
@@ -29,17 +30,35 @@
...
+      root_ca.not_before = Time.now
+      root_ca.not_after = root_ca.not_before + 2 * 365 * 24 * 60 * 60 # 2 years validity
+      root_ca.sign(root_key, OpenSSL::Digest::SHA256.new)
+      root_ca
+    end
+
+    root_key = OpenSSL::PKey::RSA.new(2048)
+    file = File.new(key, "wb")
+    file.write(root_key)
+    file.close
```
Comment (misaka) [inline config/puma.rb]: Not a blocker, but more Ruby-like alternatives:

```
File.open(key, 'wb') do |file|
  file.write(root_key)
end
```

or

```
File.write(key, root_key, 'wb')
```

Comment (tvararu) [reply config/puma.rb]: Done

File: config/puma.rb (hunk 4 above)
Comment (tvararu) [reply config/puma.rb]: Done

**Comment by tvararu** [comment]
I've tested this locally by pulling a separate repo. There's difficulties with Safari sometimes caching the certificate unreasonably past browser restarts and keychain adds, but from what I can tell, it works.

**Pull request merged**